hybrid/compartment_examples/comp_setup/shared.S

```diff
@@ -2,9 +2,11 @@
 // SPDX-License-Identifier: MIT OR Apache-2.0
 
 #define COMP_COUNT 1
-#define COMP_SIZE 48
-#define COMP_OFFSET_DDC 16
-#define COMP_OFFSET_PCC 32
+#define COMP_SIZE 80
```

Comment:
Is it possible to import the header file from the `.S` file and use `sizeof`? [This might be a very stupid question that everyone knows is impossible!]

Reply:
I'm thinking that might complicate things, and it's fine to have a bit of hard labour here? Just to clarify, you are suggesting deriving these offsets from the structure of `struct comp`, rather than having hardcoded values, is that right?

Reply:
I think at some point we'll change the `struct`, forget to change this `#define`, and then have no idea why things crash in weird ways :)

If we can use `sizeof` from assembly, that would be easiest, but my guess is it's not possible... Maybe there's some other cunning way of keeping the two in sync -- or, at least, putting the size `#define` near the struct so we're more likely to remember to change it?

Reply:
I am assuming here we can't, since we're not compiling C code, but just `asm` (@jacobbramley can intervene if I'm incorrect). And when I had the `#define`s in C code, the compiler complained, presumably as they are compiled individually. Also, since these are macros, won't that mean they will just be replaced in `asm` code with what they are defined to? I have yet to try, but I can play around a bit and see. What we might need is have inline `asm` alongside C code, which I think will be a bother...

Reply:
Indeed, it might not be possible! Let's see if @jacobbramley has any ideas; if not, we should give up and move on.

Reply:
Neither the assembler nor the C pre-processor understand C struct layouts, so there's no easy way (that I know of). I'm sure something _clever_ is possible, but probably not ideal for an example (unless it's also very clear).

The most robust method that I'm aware of is to `#define` it to a literal (like here) and `(static_)assert` that it matches `sizeof` in a C file somewhere. Likewise with `offsetof` for the fields.

Reply:
> The most robust method that I'm aware of is to #define it to a literal (like here) and (static_)assert that it matches sizeof in a C file somewhere. Likewise with offsetof for the fields.

Aha, that's clever. So that would be in a `.h` file that can be `#include`d in the `.S` file?

Reply:
Something like what I did in https://github.com/capablevms/cheri-examples/pull/52/commits/885b2af1214741666e428d41b99b2a21fc9ddc0c ?

Reply:
Yes, I think that's what Jacob's suggesting, although these might as well be `static_assert`s assuming we support C11?

Reply:
Updated in https://github.com/capablevms/cheri-examples/pull/52/commits/4c4a6112c827294f8eebfe812c087a46bf98e379, with the "it works on my setup" tag, so I assume it should work otherwise.